File: transcoder.py
#!/usr/bin/python3

import os
import sys
import sqlite3
from sqlite3 import Error
from pathlib import Path
import config

def main():
    env = os.environ.copy()

    PLEX_TRANSCODER = config.PLEX_PATH + os.sep + "Plex Transcoder_org"
    PLEX_DATABASE_PATH = config.PLEX_LIBRARY_PATH + "/Application Support/Plex Media Server/Plug-in Support/Databases/com.plexapp.plugins.library.db"

    try:
        conn = sqlite3.connect(PLEX_DATABASE_PATH)
        cur = conn.cursor()

        streams = 0
        maps = 0
        path = None
        audio_path = None
        audio_index = None
        audio_codec = None
        args_iter = iter(sys.argv)
        video_map = None
        no_accurate_seek = False
        for arg in args_iter:
            if arg == "-no_accurate_seek":
                no_accurate_seek = True
            elif arg.find("-codec:") == 0:
                codec = next(args_iter)
                stream = arg[arg.find(":")+1:]
                print("stream", stream)
                try:
                    stream = int(stream)
                except ValueError:
                    stream = 0

                if stream >= 1000:
                    audio_codec = codec

            elif arg == "-i":
                if streams == 0:
                    path = next(args_iter)
                streams = streams + 1

            elif arg == "-map":
                maps = maps + 1
                map = next(args_iter)

                if map.find(":") == -1:
                    continue

                tmp = map.split(":")
                try:
                    index = int(tmp[1])
                except ValueError:
                    index = 0

                if index < 1000:
                    continue

                print(streams)
                print("path: %s" % path)
                if path.startswith("http://127.0.0.1:32400/library/parts/"):
                    media_part_id = path.split('/')[5]
                    print(media_part_id)
                    cur.execute('SELECT `media_item_id` FROM `media_parts` WHERE `id` = ? LIMIT 1', (media_part_id,))
                    (media_item_id,) = cur.fetchone()
                else:
                    cur.execute('SELECT `id`, `media_item_id` FROM `media_parts` WHERE `file` = ? LIMIT 1', (path,))
                    (media_part_id, media_item_id) = cur.fetchone()

                print("media_part_id: %s, media_item_id: %s" % (media_part_id, media_item_id))

                cur.execute('SELECT `url`, `url_index` FROM `media_streams` WHERE `media_part_id` = ? AND `media_item_id` = ? AND `index` = ? LIMIT 1', (media_part_id, media_item_id, index,))
                (url, url_index) = cur.fetchone()
                audio_path = url[7:]
                audio_index = index - 1000

                if url_index != None:
                    audio_index = url_index

            elif arg == "-filter_complex":
                filter_complex = next(args_iter)
                print("filter_complex: %s" % filter_complex)

                j = filter_complex.find(']', 3)

                map = filter_complex[1:j]
                print("map: %s" % map)
                if map == "0:0" and filter_complex.find('scale=') > 0:
                    k = filter_complex.rfind('[', j)
                    video_map = filter_complex[k:filter_complex.find(']', k+1)+1]
                    print("skip: %s" % video_map)
                    continue

                if map.find(":") == -1:
                    continue

                tmp = map.split(":")
                try:
                    index = int(tmp[1])
                except ValueError:
                    index = 0

                audio_index = index

                if index >= 1000:
                    print(streams)

                    print("path: %s" % path)
                    if path.startswith("http://127.0.0.1:32400/library/parts/"):
                        media_part_id = path.split('/')[5]
                        print(media_part_id)
                        cur.execute('SELECT `media_item_id` FROM `media_parts` WHERE `id` = ? LIMIT 1', (media_part_id,))
                        (media_item_id,) = cur.fetchone()
                    else:
                        cur.execute('SELECT `id`, `media_item_id` FROM `media_parts` WHERE `file` = ? LIMIT 1', (path,))
                        (media_part_id, media_item_id) = cur.fetchone()

                    print("media_part_id: %s, media_item_id: %s" % (media_part_id, media_item_id))

                    cur.execute('SELECT `url`, `url_index` FROM `media_streams` WHERE `media_part_id` = ? AND `media_item_id` = ? AND `index` = ? LIMIT 1', (media_part_id, media_item_id, index,))
                    (url, url_index) = cur.fetchone()
                    audio_path = url[7:]
                    audio_index = index - 1000

                    if url_index != None:
                        audio_index = url_index

        print("maps: %d" % maps)
        if maps < 2:
            print("Probably audio streaming")
            print(sys.argv)
            os.execve(PLEX_TRANSCODER, sys.argv, env)

        #if audio_path == None:
        #    print("Not custom audio")
        #    print(sys.argv)
        #    os.execve(PLEX_TRANSCODER, sys.argv, env)

        args = []
        args_iter = iter(sys.argv)
        i = 0
        R = False

        ss = None

        for arg in args_iter:
            if arg.find("-codec:") == 0:
               codec = next(args_iter)
               stream = arg[arg.find(":")+1:]
               try:
                   stream = int(stream)
               except ValueError:
                   stream = -1

               if config.DISABLE_TRANSCODING == True and R == True and stream == 0:
                   codec = "copy"

               if stream < 1000:
                   args.extend([arg, codec])

            elif arg == "-ss":
                ss = next(args_iter)
                args.extend([arg, ss])

            elif arg == "-i":
                if video_map != None:
                    R = True
                args.extend(["-i", next(args_iter)])
                i = i + 1
                if audio_path != None and i == streams:
                    if audio_codec != None:
                        args.extend(["-codec:%d" % audio_index, audio_codec])
                    if ss != None:
                        args.extend(["-ss", ss])
                    #if no_accurate_seek == True:
                    #    args.append("-no_accurate_seek")
                    args.extend(["-probesize", "10000000", "-i", audio_path])

            elif arg == "-map":
                map = next(args_iter)

                print("map: %s, video_map: %s" % (map, video_map,))
                if config.DISABLE_TRANSCODING == True and map == video_map:
                    map = "0:0"

                if map.find(":") != -1:
                    tmp = map.split(":")
                    try:
                        index = int(tmp[1])
                    except ValueError:
                        index = 0
                    if index >= 1000:
                        map = "%d:%d" % (streams, audio_index,)

                args.append(arg)
                args.append(map)

            elif arg == "-filter_complex":
                filter_complex = next(args_iter)
                j = filter_complex.find(']', 3)

                map = filter_complex[1:j]
                print("map: %s" % map)
                if config.DISABLE_TRANSCODING == True and map == "0:0" and filter_complex.find('scale=') > 0:
                    print("skip: %s" % filter_complex)
                    continue

                print(audio_path)

                if audio_path != None and map.find(":") != -1:
                    tmp = map.split(":")
                    try:
                        index = int(tmp[1])
                    except ValueError:
                        index = 0

                    print("tmp: %s" % tmp)
                    if index >= 1000:
                        filter_complex = "[%d:%d]%s" % (streams, audio_index, filter_complex[j+1:],)

                args.extend([arg, filter_complex])

            elif config.DISABLE_TRANSCODING == True and ( arg == "-crf:0" or arg == "-maxrate:0" or arg == "-bufsize:0" or arg == "-r:0" or arg == "-preset:0" or arg == "-level:0" or arg == "-x264opts:0" or arg == "-force_key_frames:0" ):
                next(args_iter)
                continue

            elif arg == "-loglevel": #or arg == "-loglevel_plex":
                args.extend([arg, "verbose"])
                next(args_iter)
                continue

            else:
                args.append(arg)
                continue

        args[0] = 'Plex Transcoder_org'

        print("Custom audio detected")
        print(sys.argv)
        print("new:")
        print(args)

        conn.commit()
        conn.close()

        if config.DEBUG == False:
            os.execve(PLEX_TRANSCODER, args, env)
    except Error as e:
        print(e)
        sys.exit(1)


if __name__ == '__main__':
    main()

File: transcoder.go
package main

import (
	"bufio"
	"database/sql"
	"fmt"
	_ "github.com/mattn/go-sqlite3"
	"io/ioutil"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
	"strconv"
	"strings"
	"syscall"
)

func getDBPath() string {
	var p string

	// Docker
	p = "/config/Library/Application Support/Plex Media Server/Plug-in Support/Databases/com.plexapp.plugins.library.db"
	if _, err := os.Stat(p); err == nil {
		return p
	}

	// Debian, Fedora, CentOS, Ubuntu
	p = "/var/lib/plexmediaserver/Library/Application Support/Plex Media Server/Plug-in Support/Databases/com.plexapp.plugins.library.db"
	if _, err := os.Stat(p); err == nil {
		return p
	}

	// FreeBSD
	p = "/usr/local/plexdata/Plex Media Server/Plug-in Support/Databases/com.plexapp.plugins.library.db"
	if _, err := os.Stat(p); err == nil {
		return p
	}

	// ReadyNAS
	p = "/apps/plexmediaserver/MediaLibrary/Plex Media Server/Plug-in Support/Databases/com.plexapp.plugins.library.db"
	if _, err := os.Stat(p); err == nil {
		return p
	}

	home, err := os.UserHomeDir()
	if err == nil {
		// Windows
		if runtime.GOOS == "windows" {
			p = home + "\\AppData\\Local\\Plex Media Server\\Plug-in Support\\Databases\\com.plexapp.plugins.library.db"
			if _, err := os.Stat(p); err == nil {
				return p
			}
		}

		// macOS
		p = home + "/Library/Application Support/Plex Media Server/Plug-in Support/Databases/com.plexapp.plugins.library.db"
		if _, err := os.Stat(p); err == nil {
			return p
		}
	}

	ex, err := os.Executable()
	if err != nil {
		panic(err)
	}
	exPath := filepath.Dir(ex)
	return exPath + "/com.plexapp.plugins.library.db"
}

func runTranscoder(args []string) {
	if runtime.GOOS == "windows" {
		ext := filepath.Ext(args[0])

		path := args[0]
		orgFilename := path[0:len(args[0])-len(ext)] + "_org.exe"
		cmd := exec.Command(orgFilename, args[1:]...)

		go func() {
			// wait until the parent dies and bufio closes the stdin
			_, _ = ioutil.ReadAll(os.Stdin)
			if cmd.Process != nil {
				_ = cmd.Process.Kill()
			}
		}()

		_, err := cmd.Output()
		if err != nil {
			log.Fatal(err)
		}
	} else {
		err := syscall.Exec(args[0]+"_org", args, os.Environ())
		if err != nil {
			log.Fatal(err)
		}
	}
}

func main() {
	f, err := os.OpenFile(os.TempDir()+"/plex-custom-audio.log", os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
	if err != nil {
		log.Fatalf("error opening log file: %v", err)
	}
	defer f.Close()

	log.SetOutput(f)
	log.SetFlags(log.LstdFlags | log.Lshortfile)

	if runtime.GOOS == "windows" {
		child := false
		for i := 1; i < len(os.Args); i++ {
			arg := os.Args[i]
			if arg == "-child" {
				child = true
				break
			}
		}
		if !child {
			//start new child for watching this process
			args := append(os.Args[1:], "-child")
			cmd := exec.Command(os.Args[0], args...)

			//this is important, bufio will close after the parent exits,
			// unlike os.Stdin which screws up, at least on linux
			cmd.Stdin = bufio.NewReader(os.Stdin)

			cmd.Start()
			_, _ = cmd.Process.Wait()
			return
		}
	}

	log.Println("Args:")
	log.Println(os.Args)

	db, err := sql.Open("sqlite3", fmt.Sprintf("file:%s?mode=rw", getDBPath()))
	if err != nil {
		log.Fatal(err, "You can add support for your configuration by creating link to com.plexapp.plugins.library.db in the same directory as this application")
	}
	defer db.Close()

	get_media_by_id_stmt, err := db.Prepare("SELECT `media_item_id` FROM `media_parts` WHERE `id` = ? LIMIT 1")
	if err != nil {
		log.Fatal(err)
	}
	defer get_media_by_id_stmt.Close()

	get_media_by_file_stmt, err := db.Prepare("SELECT `id`, `media_item_id` FROM `media_parts` WHERE `file` = ? LIMIT 1")
	if err != nil {
		log.Fatal(err)
	}
	defer get_media_by_file_stmt.Close()

	get_media_stream_url_stmt, err := db.Prepare("SELECT `url`, `url_index`, `codec` FROM `media_streams` WHERE `media_part_id` = ? AND `media_item_id` = ? AND `index` = ? LIMIT 1")
	if err != nil {
		log.Fatal(err)
	}
	defer get_media_stream_url_stmt.Close()

	getInfo := func(path string, index int) (audioPath string, audioIndex int, audioCodec string) {
		log.Println("path:", path)
		var media_part_id int
		var media_item_id int
		if strings.HasPrefix(path, "http://127.0.0.1:32400/library/parts/") {
			media_part_id, err := strconv.Atoi(strings.Split(path, "/")[5])
			if err != nil {
				log.Fatal(err)
			}

			err = get_media_by_id_stmt.QueryRow(media_part_id).Scan(&media_item_id)
		} else {
			err = get_media_by_file_stmt.QueryRow(path).Scan(&media_part_id, &media_item_id)
		}
		if err != nil {
			log.Fatal(err)
		}

		log.Printf("media_part_id: %d, media_item_id: %d\n", media_part_id, media_item_id)

		var url string
		var url_index int
		var codec string
		get_media_stream_url_stmt.QueryRow(media_part_id, media_item_id, index).Scan(&url, &url_index, &codec)
		audioPath = url[7:]
		//audioIndex = index - 1000
		audioIndex = url_index
		audioCodec = codec

		return audioPath, audioIndex, audioCodec
	}

	inputCounter := 0
	mapCounter := 0
	path := ""
	audioPath := ""
	audioIndex := -1
	audioCodec := ""
	videoMap := ""
	hasNoAccurateSeek := false
	seek := ""

	for i := 1; i < len(os.Args); i++ {
		arg := os.Args[i]
		if arg == "-no-accurate_seek" {
			hasNoAccurateSeek = true
			continue
		}
		if arg == "-ss" {
			i++
			seek = os.Args[i]
			continue
		}
		if strings.HasPrefix(arg, "-codec:") {
			i++
			streamIndex, err := strconv.Atoi(arg[7:])
			if err == nil && streamIndex >= 1000 {
				continue
			}
			//audioCodec = os.Args[i]
			//log.Printf("audioCodec: %s\n", audioCodec)
			continue
		}
		if arg == "-i" {
			i++
			if inputCounter == 0 {
				path = os.Args[i]
			}
			inputCounter++
			continue
		}
		if arg == "-map" {
			i++
			streamMap := os.Args[i]
			mapCounter++

			if strings.IndexByte(streamMap, ':') == -1 {
				continue
			}

			streamIndex, err := strconv.Atoi(strings.Split(streamMap, ":")[1])
			if err != nil {
				continue
			}

			if streamIndex < 1000 {
				log.Printf("skip: inputCounter: %d, path: %s, index: %s\n", inputCounter, path, streamIndex)
				continue
			}

			audioPath, audioIndex, audioCodec = getInfo(path, streamIndex)
			log.Printf("inputCounter: %d, audioPath: %s, audioIndex: %d\n", inputCounter, audioPath, audioIndex)
			continue
		}
		if arg == "-filter_complex" {
			i++
			filter_complex := os.Args[i]
			log.Printf("filter_complex: %s\n", filter_complex)

			j := strings.IndexByte(filter_complex, ']')
			streamMap := filter_complex[1:j]
			log.Printf("filter_complex map: %s\n", streamMap)

			if strings.IndexByte(filter_complex, ':') == -1 {
				continue
			}

			streamIndex, err := strconv.Atoi(strings.Split(streamMap, ":")[1])
			if err != nil {
				// index = 0
				log.Fatal(err)
			}
			audioIndex = streamIndex

			if streamIndex >= 1000 {
				log.Println("Path:", path)
				audioPath, audioIndex, audioCodec = getInfo(path, streamIndex)
			}
			log.Printf("audioPath: %s, audioIndex: %d, audioCodec: %s\n", audioPath, audioIndex, audioCodec)
			continue
		}
	}

	args := []string{os.Args[0]}

	if mapCounter < 2 || len(audioPath) == 0 {
		log.Println("Probably audio streaming")

		for i := 1; i < len(os.Args); i++ {
			arg := os.Args[i]

			if arg == "-child" {
				continue
			}
			// default
			args = append(args, arg)
		}
		runTranscoder(args)
		return
	}

	log.Printf("audioPath: %s, audioIndex: %d\n", audioPath, audioIndex)

	currentInputIdx := 0

	for i := 1; i < len(os.Args); i++ {
		arg := os.Args[i]

		if arg == "-child" {
			continue
		}

		// Add -codec only if it's for valid input
		if strings.HasPrefix(arg, "-codec:") {
			i++
			index, err := strconv.Atoi(arg[7:])
			if err != nil || index < 1000 {
				args = append(args, arg, os.Args[i])
			}
			continue
		}

		// Input flag
		if arg == "-i" {
			i++
			currentInputIdx++

			// Append flags as they are
			args = append(args, arg, os.Args[i])

			// If it's last not last input - skip
			if currentInputIdx != inputCounter {
				continue
			}

			// Add codec flag
			if audioCodec != "" {
				args = append(args, fmt.Sprintf("-codec:%d", audioIndex), audioCodec)
			}

			// Add seeking flag
			if seek != "" {
				args = append(args, "-ss", seek)
			}

			// Add -no_accurate_seek
			if hasNoAccurateSeek {
				// With this switch, it's faster, but sometimes at the beginning for a few seconds there is no audio
				// args = append(args, "-no_accurate_seek")
			}

			// Add new input with audio file
			args = append(args, "-analyzeduration", "20000000", "-probesize", "20000000", "-i", audioPath)
			continue
		}

		// Audio identifier could be specified in this flag
		if arg == "-map" {
			i++
			streamMap := os.Args[i]

			log.Printf("streamMap: %s, videoMap: %s\n", streamMap, videoMap)
			if strings.IndexByte(streamMap, ':') != -1 {
				index, err := strconv.Atoi(strings.Split(streamMap, ":")[1])
				// Replace invalid stream identifier for audio with working one
				if err == nil && index >= 1000 {
					streamMap = fmt.Sprintf("%d:%d", inputCounter, audioIndex)
				}
			}

			args = append(args, arg, streamMap)
			continue
		}

		// Audio identifier could be specified in this flag
		if arg == "-filter_complex" {
			i++
			filter_complex := os.Args[i]
			log.Printf("filter_complex: %s\n", filter_complex)

			j := strings.IndexByte(filter_complex, ']')
			streamMap := filter_complex[1:j]
			log.Printf("filter_complex map: %s\n", streamMap)

			if audioPath != "" && strings.IndexByte(filter_complex, ':') != -1 {
				index, err := strconv.Atoi(strings.Split(streamMap, ":")[1])
				if err != nil {
					// index = 0
					log.Fatal(err)
				}
				if index >= 1000 {
					// Replace invalid stream identifier for audio with working one
					filter_complex = fmt.Sprintf("[%d:%d]%s", inputCounter, audioIndex, filter_complex[j+1:])
				}
			}

			args = append(args, arg, filter_complex)
			continue
		}
		// default
		args = append(args, arg)
	}

	log.Println("Custom audio detected. New args:")
	log.Println(args)

	runTranscoder(args)
}
